### PR TITLE
add huawei cloud kms wrapper

### PR DESCRIPTION
**Pull request by zengchen1024**

### CODE REVIEW

**Comment by zengchen1024** [comment]
@jefferai Could you take a look at this pr. Thanks very much!!!

File: wrappers/huaweicloudkms/huaweicloudkms.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+	"github.com/huaweicloud/golangsdk"
+	huaweisdk "github.com/huaweicloud/golangsdk/openstack"
+	kmsKeys "github.com/huaweicloud/golangsdk/openstack/kms/v1/keys"
+)
+
+// These constants contain the accepted env vars; the Vault one is for backwards compat
+const (
+	EnvHuaweiCloudKMSWrapperKeyID   = "HUAWEICLOUDKMS_WRAPPER_KEY_ID"
+	EnvVaultHuaweiCloudKMSSealKeyID = "VAULT_HUAWEICLOUDKMS_SEAL_KEY_ID"
```
Comment (jefferai) [inline wrappers/huaweicloudkms/huaweicloudkms.go]: You don't need the VAULT one as this never existed directly in Vault.

File: wrappers/huaweicloudkms/huaweicloudkms.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+func (k *Wrapper) SetConfig(config map[string]string) (map[string]string, error) {
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	// Check and set KeyID
+	keyID, err := getConfig(
+		"kms_key_id",
+		os.Getenv(EnvHuaweiCloudKMSWrapperKeyID),
+		os.Getenv(EnvVaultHuaweiCloudKMSSealKeyID),
```
Comment (jefferai) [inline wrappers/huaweicloudkms/huaweicloudkms.go]: Same comment

File: wrappers/huaweicloudkms/huaweicloudkms.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+func getConfig(name string, values ...string) (string, error) {
+	for _, v := range values {
+		if "" != v {
+			return v, nil
+		}
+	}
+
+	return "", fmt.Errorf("'%v' not found for HuaweiCloud KMS wrapper configuration", name)
```
Comment (jefferai) [inline wrappers/huaweicloudkms/huaweicloudkms.go]: nitpick: this should be `%s`, not `%v`

File: wrappers/huaweicloudkms/huaweicloudkms.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+const (
+	EnvHuaweiCloudKMSWrapperKeyID   = "HUAWEICLOUDKMS_WRAPPER_KEY_ID"
+	EnvVaultHuaweiCloudKMSSealKeyID = "VAULT_HUAWEICLOUDKMS_SEAL_KEY_ID"
+)
+
+// Wrapper is a Wrapper that uses HuaweiCloud's KMS
+type Wrapper struct {
+	client       kmsClient
+	keyID        string
+	currentKeyID *atomic.Value
```
Comment (jefferai) [inline wrappers/huaweicloudkms/huaweicloudkms.go]: I'm not sure about this currentKeyID mechanism. If this value can be changed on every encrypt operation, then it seems like any time you do a KeyID() call, this value might have been overwritten multiple times. In other words, I'm not sure it's meaningful to display the current actual value as opposed to the original alias.

Comment (zengchen1024) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Other wrappers also change the `currentKeyID` every time the encrypt operation happens. See [aws](https://github.com/hashicorp/go-kms-wrapping/blob/master/wrappers/awskms/awskms.go#L187-L194). Actually, the key id will not change for huawei cloud kms encryption operation. In other words, the input key id is same as the one the encrypt api returns.

File: wrappers/huaweicloudkms/huaweicloudkms.go (hunk 1 above)
Comment (zengchen1024) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Done

File: wrappers/huaweicloudkms/huaweicloudkms.go (hunk 2 above)
Comment (zengchen1024) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Done

File: wrappers/huaweicloudkms/huaweicloudkms.go (hunk 3 above)
Comment (zengchen1024) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Done

**Comment by zengchen1024** [comment]
@jefferai Could you take a look again. Very appreciate for your reviewing!

File: wrappers/huaweicloudkms/huaweicloudkms.go (hunk 4 above)
Comment (jefferai) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: In that case, why not just remove this mechanism entirely and return the input key id?

Comment (jefferai) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Never mind, this is fine (see https://github.com/hashicorp/go-kms-wrapping/pull/8/files)

**Comment by jefferai** [comment]
LGTM, thanks!

File: wrappers/huaweicloudkms/huaweicloudkms.go (hunk 4 above)
Comment (zengchen1024) [reply wrappers/huaweicloudkms/huaweicloudkms.go]: Thanks!

**Pull request merged**